Utilities/rANS/src/SymbolStatistics.cxx

```diff
@@ -39,60 +39,48 @@ void SymbolStatistics::rescaleToNBits(size_t bits)
 
   size_t cumulatedFrequencies = mCumulativeFrequencyTable.back();
 
+  std::vector<uint32_t> sortIdx;
+  sortIdx.reserve(mNUsedAlphabetSymbols);
+
   // resample distribution based on cumulative frequencies_
-  for (size_t i = 1; i <= mFrequencyTable.size(); i++)
-    mCumulativeFrequencyTable[i] =
-      (static_cast<uint64_t>(newCumulatedFrequency) *
-       mCumulativeFrequencyTable[i]) /
-      cumulatedFrequencies;
-
-  // if we nuked any non-0 frequency symbol to 0, we need to steal
-  // the range to make the frequency nonzero from elsewhere.
-  //
-  // this is not at all optimal, i'm just doing the first thing that comes to
-  // mind.
   for (size_t i = 0; i < mFrequencyTable.size(); i++) {
-    if (mFrequencyTable[i] &&
-        mCumulativeFrequencyTable[i + 1] == mCumulativeFrequencyTable[i]) {
-      // symbol i was set to zero freq
-
-      // find best symbol to steal frequency from (try to steal from low-freq
-      // ones)
-      std::pair<size_t, size_t> stealFromEntry{mFrequencyTable.size(), ~0u};
-      for (size_t j = 0; j < mFrequencyTable.size(); j++) {
-        uint32_t frequency =
-          mCumulativeFrequencyTable[j + 1] - mCumulativeFrequencyTable[j];
-        if (frequency > 1 && frequency < stealFromEntry.second) {
-          stealFromEntry.second = frequency;
-          stealFromEntry.first = j;
-        }
-      }
-      assert(stealFromEntry.first != mFrequencyTable.size());
-
-      // and steal from it!
-      if (stealFromEntry.first < i) {
-        for (size_t j = stealFromEntry.first + 1; j <= i; j++)
-          mCumulativeFrequencyTable[j]--;
-      } else {
-        assert(stealFromEntry.first > i);
-        for (size_t j = i + 1; j <= stealFromEntry.first; j++)
-          mCumulativeFrequencyTable[j]++;
-      }
+    if (mFrequencyTable[i]) {
+      sortIdx.push_back(i); // we will sort only those memorize only those entries which can be used
     }
   }
 
+  std::sort(sortIdx.begin(), sortIdx.end(), [this](uint32_t i, uint32_t j) { return this->getCumulativeFrequency(i) < this->getCumulativeFrequency(j); });
+  size_t need_shift = 0;
+  for (size_t i = 0; i < sortIdx.size(); i++) {
+    if (static_cast<uint64_t>(getCumulativeFrequency(sortIdx[i])) * (newCumulatedFrequency - need_shift) / cumulatedFrequencies >= 1) {
+      break;
+    }
+    need_shift++;
+  }
+
+  size_t shift = 0;
+  for (size_t i = 0; i < mFrequencyTable.size(); i++) {
+    if (static_cast<uint64_t>(getCumulativeFrequency(i)) * (newCumulatedFrequency - need_shift) / cumulatedFrequencies < 1) {
```

Comment:
I think one should check ``(static_cast<uint64_t>(newCumulatedFrequency) * mCumulativeFrequencyTable[[sortIdx[i]+1]) / cumulatedFrequencies > (static_cast<uint64_t>(newCumulatedFrequency) * mCumulativeFrequencyTable[sortIdx[i]]) / cumulatedFrequencies``, while taking the difference ``getCumulativeFrequency(sortIdx[i])`` then renormalizing, the int. arithmetic may fail.

Reply:
This will certainly not work. you need that `static_cast<uint64_t>(getFrequency(sortIdx[i])) * (newCumulatedFrequency - need_shift) / cumulatedFrequencies >= 1`, your check above misses the need_shift, and if I add it, might change its results after the need-shift has changed.

Reply:
indeed, look ok now, though I don't understand why checking int-rescaled difference (i.e. getFrequency) yields the same number of shifts as the difference of int-rescaled cumulative freqs...
Anyway, I think this can be merged.